### ISSUE TITLE
add smarter comparison when intersecting files

### ISSUE THREAD

**Issue by elsherbini** [swo/caravan#5]
I often have already demultiplexed paired end reads from the BMC. So I don't need to run the demultiplex file, but i still need to intersect the forward and reverse fastq file.

**Comment by elsherbini** [comment]
also it occurs to me that intersecting could be agnostic to which files you are giving it. just a single `van.py instersect [arbitrary list of files]` and it could figure out if they were fastq's or tsv's and do the right thing(tm).

**Comment by swo** [comment]
Does "intersect" mean "merge reads"? Does the BMC give you fastq's that don't, uh, "align" with each other (i.e., give you files where line _n_ of the forward fastq relates is not _related_ to line _n_ of the reverse)?

**Comment by elsherbini** [comment]
no, intersect before merging, but after trimming/demultiplexing, so some reads have been thrown out because they didn't make the cut in previous steps.

**Comment by swo** [comment]
I think I got it: the idea is to have an `intersect` functionality that could look at...
- two fastq's and throw out the sequences that don't have "mates"
- a fastq and a tsv, and do the same
- two fastq's and a tsv
- I guess _any_ combination of fastq's, fasta's, or tsv's and find the header/first field that have matching names.

Did I get it right?

Maybe it would be nice if `convert` could create a mapping file while it converts (e.g., if you're already demultiplexed, it's annoying to concatenate the fastq's and make that mapping separately). Maybe even better would be to have `convert` put the sample name right into the fastq header so you can avoid having the redundant mapping file.

**Comment by elsherbini** [comment]
Ooo, looks like acc221b9bcdc7af6135155cc854bf3d72dadb601 fixes this?

**Comment by swo** [comment]
Yes, it allows for arbitrary intersecting of fastq's, fasta's, and tsv's, so long as the numbering scheme follows `read1234`, i.e., `@read1234` for a fastq, `>read1234` for a fasta, and `read1234\twhatever` for a tsv.

I have trouble with figuring out how to intersect arbitrary files, since the header lines in regular fastq's don't seem to have an obvious order, so I don't know which file should be incremented to look for the match. Any thoughts here?

**Comment by elsherbini** [comment]
I think if you don't sort them somehow you are out of luck. Could either sort the headers the way you do by renaming them all to be ordered, or load all the headers into memory and assign an index to them. I think the current workflow of requiring people rename the headers to be implicitly sorted is fine though.

**Comment by swo** [comment]
Word; I think the disadvantageous thing in that pre-indexed approach is that you have to read through the file twice. I like the idea of being able to run all the data through single-pass pipes.

We could generalize a little bit by doing some kind of smart comparison between entries? I'm thinking that `read1234` is a little restrictive when it could be smart to understand that `foo;1` should come before `foo;2` and `aaa` should come before `zzz`. I'll edit this issue to that effect.

**Comment by elsherbini** [comment]
Natsort (https://pypi.python.org/pypi/natsort) is pretty intelligent. So natsort the list of current headers, increment the one that is smallest, rinse,repeat until they all match?
